### PR TITLE
fix: Button disabled style with link or href case

### LINKED ISSUE #42997: button type = 'link' 并且禁用的情况下 样式问题
### Reproduction link

https://codesandbox.io/s/antd-reproduction-template-forked-pcwx68?file=/index.js

### Steps to reproduce

link 后面加个 href

### What is expected?

正常的link禁用样式

### What is actually happening?

正常button禁用的样式

| Environment | Info |
| --- | --- |
| antd | 5.6.1 |
| React | 18.0.0 |
| System | mac |
| Browser | 113.0.5672.126 |

### PR DESCRIPTION
**Pull request by BoyYangzai**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
- fix https://github.com/ant-design/ant-design/issues/42997

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     fix: Button disabled style with link or href case      |
| 🇨🇳 Chinese |     修复 Button 在有 link or href 属性时禁用样式错误      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### CODE REVIEW

File: components/button/button.tsx [hunk 1]
```diff
@@ -263,7 +260,9 @@ const InternalButton: React.ForwardRefRenderFunction<
     return wrapSSR(
       <a
         {...linkButtonRestProps}
-        className={classes}
+        className={classNames(classes, {
+          [`${prefixCls}-disabled`]: mergedDisabled,
```
Comment (MadCcc) [inline components/button/button.tsx:264]: 可以合到上面去吧

Comment (Wxh16144) [reply components/button/button.tsx:264]: 放在这里更好阅读一点, 清晰的知道只有 render 的不是 <button /> 元素的时候才会出现 ant-btn-disable classname

**Review by afc163 — APPROVED** [review]

**Review by afc163 — DISMISSED** [review]
看下 lint 问题

File: components/button/style/index.ts [hunk 2]
```diff
@@ -115,8 +115,12 @@ const genSharedButtonStyle: GenerateStyle<ButtonToken, CSSObject> = (token): CSS
...
 };
 
-const genHoverActiveButtonStyle = (hoverStyle: CSSObject, activeStyle: CSSObject): CSSObject => ({
-  '&:not(:disabled)': {
+const genHoverActiveButtonStyle = (
+  btnCls: string,
+  hoverStyle: CSSObject,
+  activeStyle: CSSObject,
+): CSSObject => ({
+  [`&:not(:disabled, ${btnCls}-disabled)`]: {
```
Comment (Wxh16144) [inline components/button/style/index.ts:123]:
```suggestion
  [`&:not(:disabled):not(${btnCls}-disabled)`]: {
```

Comment (Wxh16144) [reply components/button/style/index.ts:123]: ci lint 报错, 但是这里的逻辑确实应该要这样写呀

Comment (MadCcc) [reply components/button/style/index.ts:123]: cssinjs 不允许这样写，低版本浏览器不兼容

**Review by MadCcc — APPROVED** [review]

**Pull request merged**